probe_scraper/transform_probes.py

```diff
@@ -353,8 +353,20 @@ def transform_by_hash(commit_timestamps, metric_data):
         sorted_commits = sorted(iter(commits.items()),
                                 key=lambda x_y: commit_timestamps[repo_name][x_y[0]])
 
+        # Not only do we need to know when metrics were added or changed, we also
```

Comment:
Can we add a test for this change? It's not clear to me what the expected outcome here is:
- Should there be an entry in `history` with just `{"disabled": True}`?
- Should this probe stop being reported in the probe-scraper altogether?

Reply:
Yes -- if a probe is removed, it should have an entry with the commit where it was removed, and the only contents are `{"disabled": True}`.

However, I don't think it should stop being reported in the probe-scraper altogether, since we want to retain the database columns for historical data before the metric was expired (I think) -- though that kind of relates to questions of data retention and when we would go back and cull columns for expired metrics.  Regardless, if we don't make this change, the probe-scraper can't communicate downstream to any tools that a metric was removed (and thus disabled).

Reply:
I've added a test here.

Reply:
Great, thanks, this helps me understand what's happening. So the idea is that we'll remove a duplicate probe, and then it will show up as `"disabled": True` in the probe-info-service?

There's an additional hiccup, in that MSG doesn't know anything about this non-definitioned probe (a disabled one). It will try to get definition for it to fill-in the schema, and fail.

Question then: How do we decipher between probes that are _just_ disabled, and probes that were duplicates and disabled? The former should show up in the schema, but the latter should not.

If we do something fancy like "if a dupe is found in MSG, take the one that is not disabled", how do we handle the future case where dupe happens, one metric is disabled, and then sometime down the line, the other one is removed/disabled?

Reply:
Given these issues, and what we discussed in the meeting today, I think (a) we want to have *something* to start detecting this problem sooner than later, but (b) finding a good way forward to resolve these issues once they are created is going to be difficult.

Based on that, I propose:

1) I back out the changes here that actually modify the output (the insertion of `"disabled": True` entries in the history list).  That solution is perhaps half-baked.
2) Word the e-mail carefully such that it's clear all developers should reach out for help.
3) Deal with the problem of how to actually resolve any conflicts when we know we actually have them (since the solution may be somewhat case-by-case anyway).
4) Continue to tackle this problem on better fronts -- either by "fully namespacing" or build-time detection.

What do you think?

Reply:
I think what you've outlined is wise. If we're lucky we won't have to deal with this problem :)